<a href="https://colab.research.google.com/github/DhimanTarafdar/AAA/blob/main/20_practice_day.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**#PyTorch Mini Project (Dataset-Independent)**
Choose One Dataset

Students must choose one dataset from the following list and build a PyTorch ANN model.

1️⃣ MNIST Handwritten Digit Dataset

Digits 0–9 classification.

Dataset link:
https://www.kaggle.com/datasets/hojjatk/mnist-dataset

2️⃣ Fashion-MNIST Dataset

Clothing image classification.

Dataset link:
https://www.kaggle.com/datasets/zalando-research/fashionmnist

3️⃣ Iris Dataset

Flower species classification.

Dataset link:
https://www.kaggle.com/datasets/uciml/iris

4️⃣ Breast Cancer Wisconsin Dataset

Binary classification (malignant vs benign).

Dataset link:
https://www.kaggle.com/datasets/uciml/breast-cancer-wisconsin-data

5️⃣ Titanic Survival Dataset

Predict whether a passenger survived.

Dataset link:
https://www.kaggle.com/datasets/yasserh/titanic-dataset

# Topics Students Must Use

Students must apply the following PyTorch concepts in their project:

Why PyTorch

Tensors

Autograd

PyTorch Training Pipeline

torch.nn Module & torch.optim Module

Dataset and DataLoader

Building ANN using PyTorch

Model training and evaluation




# Question 1

Load the dataset you selected and perform basic preprocessing.

Explain:

number of samples

number of features

target variable

train-test split

# Write answer of Question 1

# **Step 1: Dataset Load**

In [1]:
import pandas as pd

# load dataset
df = pd.read_csv("data.csv")

# unnecessary column drop (id column usually useless)
df = df.drop(columns=["id"], errors='ignore')

# target column rename for clarity
df = df.rename(columns={"diagnosis": "target"})

df.head()

,target,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,symmetry_mean,...,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst,Unnamed: 32
0,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,NaN
1,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,NaN
2,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,NaN
3,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,NaN
4,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,NaN


# **Step 2: Basic Info**

In [8]:
print("Dataset Shape:", df.shape)
print("\nColumns:\n", df.columns)

# Drop unnecessary columns
df = df.drop(columns=["id", "Unnamed: 32"], errors='ignore')

Dataset Shape: (569, 32)

Columns:
 Index(['target', 'radius_mean', 'texture_mean', 'perimeter_mean', 'area_mean',
       'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst', 'Unnamed: 32'],
      dtype='object')


In [9]:
df

,target,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,symmetry_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,1,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.30010,0.14710,0.2419,...,25.380,17.33,184.60,2019.0,0.16220,0.66560,0.7119,0.2654,0.4601,0.11890
1,1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.08690,0.07017,0.1812,...,24.990,23.41,158.80,1956.0,0.12380,0.18660,0.2416,0.1860,0.2750,0.08902
2,1,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.19740,0.12790,0.2069,...,23.570,25.53,152.50,1709.0,0.14440,0.42450,0.4504,0.2430,0.3613,0.08758
3,1,11.42,20.38,77.58,386.1,0.14250,0.28390,0.24140,0.10520,0.2597,...,14.910,26.50,98.87,567.7,0.20980,0.86630,0.6869,0.2575,0.6638,0.17300
4,1,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.19800,0.10430,0.1809,...,22.540,16.67,152.20,1575.0,0.13740,0.20500,0.4000,0.1625,0.2364,0.07678
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
564,1,21.56,22.39,142.00,1479.0,0.11100,0.11590,0.24390,0.13890,0.1726,...,25.450,26.40,166.10,2027.0,0.14100,0.21130,0.4107,0.2216,0.2060,0.07115
565,1,20.13,28.25,131.20,1261.0,0.09780,0.10340,0.14400,0.09791,0.1752,...,23.690,38.25,155.00,1731.0,0.11660,0.19220,0.3215,0.1628,0.2572,0.06637
566,1,16.60,28.08,108.30,858.1,0.08455,0.10230,0.09251,0.05302,0.1590,...,18.980,34.12,126.70,1124.0,0.11390,0.30940,0.3403,0.1418,0.2218,0.07820
567,1,20.60,29.33,140.10,1265.0,0.11780,0.27700,0.35140,0.15200,0.2397,...,25.740,39.42,184.60,1821.0,0.16500,0.86810,0.9387,0.2650,0.4087,0.12400


# **Step 3: Target Encoding**

In [10]:
# Convert target: M = 1 (Malignant), B = 0 (Benign)
df['target'] = df['target'].map({'M': 1, 'B': 0})

# **Step 4: Features & Labels Split**

In [11]:
X = df.drop("target", axis=1).values
y = df["target"].values

# **Step 5: Train-Test Split**

In [12]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("Train Shape:", X_train.shape)
print("Test Shape:", X_test.shape)

Train Shape: (455, 30)
Test Shape: (114, 30)


# **Step 6: Normalization (VERY IMPORTANT for ANN)**

In [13]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Question 2

Convert the dataset features into PyTorch tensors.

Explain:

what a tensor is

tensor shape of your dataset

why tensors are required in PyTorch

#Write answer of Question 2

# **Step 1: NumPy → PyTorch Tensor Convert**

In [14]:
import torch

# convert to tensors
X_train_tensor = torch.tensor(X_train, dtype=torch.float32)
X_test_tensor = torch.tensor(X_test, dtype=torch.float32)

y_train_tensor = torch.tensor(y_train, dtype=torch.float32)
y_test_tensor = torch.tensor(y_test, dtype=torch.float32)

# reshape target for ANN (important)
y_train_tensor = y_train_tensor.view(-1, 1)
y_test_tensor = y_test_tensor.view(-1, 1)

In [16]:
X_train_tensor

tensor([[-1.4408, -0.4353, -1.3621,  ...,  0.9320,  2.0972,  1.8865],
        [ 1.9741,  1.7330,  2.0917,  ...,  2.6989,  1.8912,  2.4978],
        [-1.4000, -1.2496, -1.3452,  ..., -0.9702,  0.5976,  0.0579],
        ...,
        [ 0.0488, -0.5550, -0.0651,  ..., -1.2390, -0.7086, -1.2715],
        [-0.0390,  0.1021, -0.0314,  ...,  1.0500,  0.4343,  1.2134],
        [-0.5486,  0.3133, -0.6035,  ..., -0.6110, -0.3345, -0.8463]])

# **Step 2: Check Tensor Shape**

In [15]:
print("X_train_tensor shape:", X_train_tensor.shape)
print("y_train_tensor shape:", y_train_tensor.shape)

print("X_test_tensor shape:", X_test_tensor.shape)
print("y_test_tensor shape:", y_test_tensor.shape)

X_train_tensor shape: torch.Size([455, 30])
y_train_tensor shape: torch.Size([455, 1])
X_test_tensor shape: torch.Size([114, 30])
y_test_tensor shape: torch.Size([114, 1])


### What is a Tensor

A tensor is a multi-dimensional data structure used in PyTorch. It is very similar to a NumPy array but more powerful because it can run on GPU and supports automatic differentiation. In simple terms, tensor is the basic data format that PyTorch uses to store and process data.


### Tensor Shape of My Dataset

In my dataset, the shape of the tensors shows how many samples and features I have. For example, X_train has shape (455, 30), which means there are 455 samples and each sample has 30 features. The target tensor y_train has shape (455, 1), which represents the labels for each sample.


### Why Tensors are Required in PyTorch

Tensors are required in PyTorch because all operations in neural networks are performed using tensors. They allow fast computation, support GPU acceleration, and enable automatic gradient calculation using autograd. Without tensors, we cannot train or build deep learning models in PyTorch.


# Question 3

Create a simple example using Autograd:

define a tensor with requires_grad=True

perform a mathematical operation

compute gradient using .backward()

Explain why autograd is important for neural networks.

#Write answer of Question 3

# **Step 1: Simple Autograd Example**

In [17]:
import torch

# define tensor with requires_grad=True
x = torch.tensor(2.0, requires_grad=True)

# perform operation
y = x**2 + 3*x + 1

# compute gradient
y.backward()

# print gradient
print("Value of x:", x.item())
print("Gradient (dy/dx):", x.grad.item())

Value of x: 2.0
Gradient (dy/dx): 7.0


### Why Autograd is Important for Neural Networks

Autograd is very important in PyTorch because it automatically calculates gradients during training. In neural networks, we need gradients to update weights using backpropagation. Instead of manually computing derivatives, autograd tracks all operations and computes gradients automatically. This makes model training easier, faster, and less error-prone.


# Question 4

Create a custom Dataset class and use DataLoader.

Explain the role of:

Dataset

DataLoader

batch_size

shuffle

#  Write answer of Question 4

# **Step 1: Custom Dataset Class**

In [18]:
from torch.utils.data import Dataset

class BreastCancerDataset(Dataset):
    def __init__(self, X, y):
        self.X = torch.tensor(X, dtype=torch.float32)
        self.y = torch.tensor(y, dtype=torch.float32)
        self.y = self.y.view(-1, 1)

    def __len__(self):
        return len(self.X)

    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]

# **Step 2: Create Dataset Objects**

In [19]:
train_dataset = BreastCancerDataset(X_train, y_train)
test_dataset = BreastCancerDataset(X_test, y_test)

# **Step 3: DataLoader**

In [20]:
from torch.utils.data import DataLoader

train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=32, shuffle=False)

# **Step 4: Check One Batch**

In [21]:
for X_batch, y_batch in train_loader:
    print("X batch shape:", X_batch.shape)
    print("y batch shape:", y_batch.shape)
    break

X batch shape: torch.Size([32, 30])
y batch shape: torch.Size([32, 1])


# Question 5

Design an Artificial Neural Network (ANN) using torch.nn.Module.

Your model must include:

input layer

at least one hidden layer

output layer

Explain the purpose of:

nn.Linear

activation functions

loss function

# Write answer of Question 5

# Question 6

Implement a complete PyTorch training pipeline that includes:

Forward pass

Loss calculation

Backpropagation

Optimizer step

Clearing gradients

Train the model for multiple epochs.

#  Write answer of Question 6

# Question 7

Evaluate your model on the test dataset and compute:

accuracy

loss

Explain whether the model is underfitting or overfitting.

#  Write answer of Question 7

# Question 8

Modify the architecture and compare performance:

Example modifications:

increase hidden neurons

add another hidden layer

change activation function

Explain the effect on performance.

#  Write answer of Question 8